### PR TITLE
auth k8S error solution design

### PR DESCRIPTION
**Pull request by oburstein-hub**
### What does this PR do?
This PR contains solution design document for improving authn k8S interaction with auth client and secret provide

### What ticket does this PR close?
Error in first authentication in secret provider

### CODE REVIEW

File: design/authenticators/authn_k8s_error_solution.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+**Note**: This design document has been ported from the original document
+[here](https://github.com/cyberark/conjur/issues/524).
+
+# Background
+The need to write this solution design arises from the problem that was raised as part of secret provider phase II implementation.
```
Comment (sigalsax) [inline design/authenticators/authn_k8s_error_solution.md:5]: please link to the solution doc

File: design/authenticators/authn_k8s_error_solution.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+The problem is getting error upon authenticate call from secret provider to conjur. The same error appears in all K8S authenticator client usages
+
+# Issue description
+The root cause of the issue that was found appears to be that K8S auth client tries to access to cert file that is written from Conjur K8S Authn by K8S API before this cert file appears / is ready
+
+The reason for this behaviour is that the flow returns to K8S Authn Client without waiting for cert file to be fully written, and K8S Auth client proceeds with flow to read a file - and then we get an error.
+
+# Solution
+The solution should be that K8S Authn Client should wait until and event that says that cert file is ready for use will arrive - and only then it should proceed to read it
```
Comment (sigalsax) [inline design/authenticators/authn_k8s_error_solution.md:15]: waits until "an" event? or "and"

File: design/authenticators/authn_k8s_error_solution.md (hunk 1 above)
Comment (oburstein-hub) [reply design/authenticators/authn_k8s_error_solution.md:5]: please, send me a link to the solution doc and I will add it

File: design/authenticators/authn_k8s_error_solution.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+In case that the expected event does not arrive after timeout - error must be issued and the flow will proceed without reading cert file
+
+# Design
+
+![Sequence Diagram](images/authenticators_bug_sequence_diagram.jpeg)
```
Comment (sigalsax) [inline design/authenticators/authn_k8s_error_solution.md:21]: picture is not rendered

File: design/authenticators/authn_k8s_error_solution.md (hunk 2 above)
Comment (oburstein-hub) [reply design/authenticators/authn_k8s_error_solution.md:15]: fixing, 10x

File: design/authenticators/authn_k8s_error_solution.md (hunk 3 above)
Comment (oburstein-hub) [reply design/authenticators/authn_k8s_error_solution.md:21]: fixed

File: design/authenticators/authn_k8s_error_solution.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+The addition to the current design is WaitUntilWrittenOrTimeout call. It should be executed per new flag sent into the InjectClientCert command (in order to stay backward compatible)
```
Comment (sigalsax) [inline design/authenticators/authn_k8s_error_solution.md:23]: What do you mean "executed per flag"
The design is really important to go into detail so please add a few more sentences explaining the flow

File: design/authenticators/authn_k8s_error_solution.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+See flag above
+
+# Performance
+
+In case performance is an issue - flag above will be off
+
+# Affected Components
+Conjur, AuthnClient, Secret Provider
+
+# Test plain
```
Comment (sigalsax) [inline design/authenticators/authn_k8s_error_solution.md:47]: test plan*

File: design/authenticators/authn_k8s_error_solution.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+Successful K8S authn	K8S authenticator	Authenticate from secret provider	Success	No errors
+Failed K8S authn	K8S authenticator	Authenticate with wrong cert from secret provider	Failure	Error
```
Comment (sigalsax) [inline design/authenticators/authn_k8s_error_solution.md:49]: Please create your test plan in the format similar to here: https://github.com/cyberark/secrets-provider-for-k8s/blob/master/design/milestone_1_2_design_doc.md#test-plan

File: design/authenticators/authn_k8s_error_solution.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+Failed to create cert file	K8S authenticator	Authenticate from secret provider	Timeout	Error:Timeout
+
+# Logs
+
+File not found after timeout	File not found after timeout
```
Comment (sigalsax) [inline design/authenticators/authn_k8s_error_solution.md:54]: Please add log in the format similar to: https://github.com/cyberark/secrets-provider-for-k8s/blob/master/design/milestone_1_2_design_doc.md#logs 
with the associated log level

File: design/authenticators/authn_k8s_error_solution.md (hunk 1 above)
Comment (sigalsax) [reply design/authenticators/authn_k8s_error_solution.md:5]: https://github.com/cyberark/secrets-provider-for-k8s/blob/master/design/milestone_1_2_design_doc.md

File: design/authenticators/authn_k8s_error_solution.md (hunk 4 above)
Comment (InbalZilberman) [reply design/authenticators/authn_k8s_error_solution.md:23]: @oburstein-hub why is adding this wait is breaking the behavior that was before? in what sense?

File: design/authenticators/authn_k8s_error_solution.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+The problem is getting error upon authenticate call from secret provider to conjur. The same error appears in all K8S authenticator client usages
+
+# Issue description
+The root cause of the issue that was found appears to be that K8S auth client tries to access to cert file that is written from Conjur K8S Authn by K8S API before this cert file appears / is ready
+
+The reason for this behaviour is that the flow returns to K8S Authn Client without waiting for cert file to be fully written, and K8S Auth client proceeds with flow to read a file - and then we get an error.
+
+# Solution
+The solution should be that K8S Authn Client should wait until an event that says that cert file is ready for use will arrive - and only then it should proceed to read it
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply design/authenticators/authn_k8s_error_solution.md:15]: the event is just the response from the server of the inject_client_cert request
In current code - inside the implementation of inject_client_cert there is no wait until completion of writing the file

File: design/authenticators/authn_k8s_error_solution.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+In case that the expected event does not arrive after timeout - error must be issued and the flow will proceed without reading cert file
+
+# Design
+
+![Sequence Diagram](images/authenticators_bug_sequence_diagram.jpeg)
+
+The addition to the current design is WaitUntilWrittenOrTimeout call. It should be executed per new flag sent into the InjectClientCert command (in order to stay backward compatible)
+
+New env variable will be added to AuthnClient process for the flag above
```
[the remark this answers is not in the document]
Comment (oburstein-hub) [reply design/authenticators/authn_k8s_error_solution.md:25]: I think we do need env-var to put this solution to eliminate influence on irrelevant cases

File: design/authenticators/authn_k8s_error_solution.md (hunk 4 above)
Comment (oburstein-hub) [reply design/authenticators/authn_k8s_error_solution.md:23]: After a fix system will always wait for file to appear. I wanted to preserve a behaviour when it does not wait. I might be important for some other existing scenarios. If you dont see such - we can eliminate requirement of flag

**Comment by orenbm** [comment]
Closing this as it is a  bug that does not require this design

**Pull request closed without merging**